pkg/keys/sql.go

```diff
@@ -259,3 +259,35 @@ func (d sqlDecoder) DecodeTenantMetadataID(key roachpb.Key) (roachpb.TenantID, e
 	}
 	return roachpb.MakeTenantID(id), nil
 }
+
+// RewriteSpanToTenantPrefix updates the passed Span, potentially in-place, to
+// ensure the Key and EndKey have the passed tenant prefix, regardless of what
+// prior tenant prefix, if any, they had before, and returns the updated Span.
+func RewriteSpanToTenantPrefix(sp roachpb.Span, prefix roachpb.Key) (roachpb.Span, error) {
+	var err error
+	sp.Key, err = rewriteKeyToTenantPrefix(sp.Key, prefix)
+	if err != nil {
+		return sp, err
+	}
+	sp.EndKey, err = rewriteKeyToTenantPrefix(sp.EndKey, prefix)
+	return sp, err
+}
+
+func rewriteKeyToTenantPrefix(key roachpb.Key, prefix roachpb.Key) (roachpb.Key, error) {
+	if !bytes.HasPrefix(key, prefix) {
```

Comment:
I think this only works for the `Tenant -> Tenant` and `System -> Tenant` transformations.  If we want to support `Tenant -> System`, I think this wouldn't work since our current prefix will be `""` and `HasPrefix` will then always return true.

Reply:
Fixed, and made sure the test case has both tenant and non-tenant spans and is run by both tenants and non-tenants.